clientconn_test.go

```diff
@@ -820,3 +820,128 @@ func TestBackoffCancel(t *testing.T) {
 	cc.Close()
 	// Should not leak. May need -count 5000 to exercise.
 }
+
+type signalingBackoffStrategy struct {
+	ready chan struct{}
+}
+
+func (bs *signalingBackoffStrategy) Backoff(retries int) time.Duration {
+	select {
+	case <-bs.ready:
+	}
+	return time.Duration(0)
+}
+
+func TestDialCloseStateTransition(t *testing.T) {
+	defer leakcheck.Check(t)
+
+	server, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		t.Fatalf("Error while listening. Err: %v", err)
+	}
+	defer server.Close()
+	testFinished := make(chan struct{})
+	defer close(testFinished)
+	killFirstConnection := make(chan struct{})
+	killSecondConnection := make(chan struct{})
+	exitBackoff := make(chan struct{}, 1)
+	bs := signalingBackoffStrategy{ready: exitBackoff}
+
+	// Launch the server.
+	go func() {
+		// Establish a successful connection so that we enter READY. We need to get
+		// to READY so that we can get a client back for us to introspect later (as
+		// opposed to just CONNECTING).
+		conn, err := server.Accept()
+		if err != nil {
+			t.Error(err)
+			return
+		}
+		defer conn.Close()
+
+		framer := http2.NewFramer(conn, conn)
+		if err := framer.WriteSettings(http2.Setting{}); err != nil {
+			t.Errorf("Error while writing settings frame. %v", err)
+			return
+		}
+
+		select {
+		case <-testFinished:
+			return
+		case <-killFirstConnection:
+		}
+
+		// Close the conn to cause onShutdown, causing us to enter TRANSIENT FAILURE. Note that we are not in
+		// WaitForHandshake at this point because the preface was sent successfully.
+		conn.Close()
+
+		// Re-connect (without server preface), causing us to enter CONNECTING.
+		conn, err = server.Accept()
+		if err != nil {
+			t.Error(err)
+			return
+		}
+
+		// Close the conn to cause onShutdown, causing us to enter TRANSIENT FAILURE. Note that we are in
+		// WaitForHandshake at this point because the preface has not been sent yet.
+		select {
+		case <-testFinished:
+			return
+		case <-killSecondConnection:
+		}
+		conn.Close()
+	}()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
+	defer cancel()
+
+	exitBackoff <- struct{}{} // Allow the first connect to pass without backoff.
+	client, err := DialContext(ctx, server.Addr().String(), WithInsecure(), WithWaitForHandshake(), WithBlock(), withBackoff(&bs))
+	if err != nil {
+		t.Fatal(err)
+		return
+	}
+	defer client.Close()
+	// This has to happen before client.Close, or client.Close will hang forever waiting for backoff to complete.
+	defer close(exitBackoff)
+
+	if client.GetState() != connectivity.Ready {
+		t.Fatalf("expected addrconn to be in READY, was %v", client.GetState())
+	}
+
+	close(killFirstConnection)
+	time.Sleep(100 * time.Millisecond)
```

Comment:
Please use `cc.WaitForStateChange` or else this will be racy and/or sleep unnecessarily.

Reply:
So, because I now properly use withBackoff(infiniteBackoff{}), there's another problem here: the first backoff after a successful handshake is skipped and reconnection immediately happens.

That means that I have to add another Accept->close pair to the server to get to the infinite backoff (from `close -> Accept -> signal blocks` to `close -> Accept -> client connects immediately -> close -> Accept -> backoff blocks`).

So, now when I do `close(killFirstConnection)`, the state transition is: 

```
READY -> CONNECTING -> TRANSIENT FAILURE (no backoff) -> CONNECTING -> TRANSIENT FAILURE (infinite backoff)
```

This all happens _really_ fast. When I try to write code that manually follows these five transitions, my test fails 50% of the time. I added `sleep(50ms)` to just skip over all of them. It's not great, but I don't have a better idea. It worked 100% of the time when I did `-cpu 1,4 -count 1000` locally, FWIW.